File: Assets/CSScripts/Common/LoadTest.cs
using System.Threading.Tasks;
using UnityEngine;
using UnityEngine.UI;
using XiaoXu.Core;
using System.Collections.Generic;

public class LoadTest : MonoBehaviour
{
	[Header("标签")]
	public string label;

	[Header("图片")]
	public Image targetImage;
	public string imageSpriteAddress;

	[Header("敌人")]
	public string enemyName;

    [Header("实例化设置")]
	public Transform spawnParent;
	public Vector3 spawnPosition = Vector3.zero;

	private GameMain gameMain;
	private ResourceManager resourceManager;
	private List<GameObject> spawnedObjects = new List<GameObject>();

	async void Start()
	{
		gameMain = GetComponent<GameMain>();
		resourceManager = gameMain.GetComponentInChildren<ResourceManager>();

		await LoadAndSpawnByLabel();
		await LoadAndSetSprite();
		await LoadAndSpawnByName();
	}

	private async Task LoadAndSpawnByLabel()
	{
		List<GameObject> loadedPrefabs = await resourceManager.LoadAssetsByLabelAsync<GameObject>(label);
		Debug.Log($"通过标签 {label} 加载了 {loadedPrefabs.Count} 个GameObject");

		foreach (GameObject prefab in loadedPrefabs)
		{
			if (prefab != null)
			{
				GameObject instance = await resourceManager.InstantiateAsync(prefab.name,spawnPosition,Quaternion.identity);

				if (instance != null)
				{
					if (spawnParent != null)
					{
						instance.transform.SetParent(spawnParent, false);
					}

					spawnedObjects.Add(instance);
				}
			}
		}
	}
	private async Task LoadAndSpawnByName()
	{
        GameObject obj = await resourceManager.InstantiateAsync(enemyName, spawnPosition, Quaternion.identity);
	}
	private async Task LoadAndSetSprite()
	{
		Sprite loadedSprite = await resourceManager.LoadAssetAsync<Sprite>(imageSpriteAddress);

		if (loadedSprite != null && targetImage != null)
		{
			targetImage.sprite = loadedSprite;
			targetImage.preserveAspect = true; // 保持图片比例
		}
		else
		{
			Debug.LogError($"加载图片失败: {imageSpriteAddress}");
		}
	}
	void OnDestroy()
	{
		// 销毁所有实例化的对象
		foreach (GameObject obj in spawnedObjects)
		{
			if (obj != null)
			{
				resourceManager.DestroyInstance(obj);
			}
		}
		spawnedObjects.Clear();

		// 释放图片资源
		if (!string.IsNullOrEmpty(imageSpriteAddress))
		{
			resourceManager.ReleaseAsset(imageSpriteAddress);
		}

		// 释放标签资源
		resourceManager.ReleaseAssetsByLabel(label);

		Debug.Log("LoadTest 资源清理完成");
	}
}

File: Assets/lua~/Game/Test/LoadTest.lua
---@class LoadTest
LoadTest = DefineClass()

function LoadTest:Start()
    if not GameMain then
        print("GameMain为空")
    else
        print("GameMain不为空")
    end

    self:LoadAndSpawnByLabel()
    self:LoadAndSetSprite()
    self:LoadAndSpawnByName()
end
--用标签加载一组物体
function LoadTest:LoadAndSpawnByLabel()
    if not GameMain.resourceManager then
        print("警告: GameMain.resourceManager 不存在，尝试直接获取")
    end

    local assets = ResourceManager:LoadAssetsByLabelAsync(self.label)
    for _, asset in ipairs(assets) do
        if asset ~= nil then
            local instance = ResourceManager:Instantiate(asset.name, self.spawnPosition, self.spawnRotation)
            if instance ~= nil and self.spawnParent ~= nil then
                instance.transform:SetParent(self.spawnParent, false)
            end
        end
    end
end

function LoadTest:LoadAndSpawnByName()
    if string.IsNullOrEmpty(self.enemyName) then 
        return 
    end
    
    local enemy = ResourceManager:InstantiateAsync(self.enemyName, self.spawnPosition, self.spawnRotation)
    if enemy ~= nil and self.spawnParent ~= nil then
        enemy.transform:SetParent(self.spawnParent, false)
    end
end

function LoadTest:LoadAndSetSprite()
    if string.IsNullOrEmpty(self.imageSpriteAddress) or self.targetImage == nil then return end
    
    local sprite = ResourceManager:LoadAssetAsync(self.imageSpriteAddress)
    if sprite ~= nil then
        self.targetImage.sprite = sprite
        self.targetImage.preserveAspect = true
        print("图片加载成功!")
    else
        print("图片加载失败!")
    end
end

function LoadTest:OnDestroy()
    if not string.IsNullOrEmpty(self.imageSpriteAddress) then
        ResourceManager:ReleaseAsset(self.imageSpriteAddress)
    end
    
    if not string.IsNullOrEmpty(self.label) then
        ResourceManager:ReleaseAssetsByLabel(self.label)
    end
end

return LoadTest